amp-pwa-reader/src/js/Backend.js

```diff
@@ -121,7 +125,23 @@ class Backend {
       this._imageRatio = featuredImage.getAttribute('height') / featuredImage.getAttribute('width');
       featuredImage.remove();
     }
+  }
+
+// Inject CSS into the <style amp-custom> tag.
+// Note that the inlineCSS property gets added to this object by inline.css.js.
+  injectCSS(html) {
+    const styleTag = '<style amp-custom>';
+
+    return html.replace(styleTag, styleTag + ' ' + this.hideElementsCSS() + this.inlineCSS);
+  }
 
+// Given an array of selectors, create CSS rules that hide each of those - and insert it into AMP HTML.
+// This is the equivalent of sanitize() for the streaming case.
+// TODO: We don't grab the metadata that sanitize() grabs, which could matter someday
+  hideElementsCSS(html) {
```

Comment:
TODO: Would be more elegant to simply remove these elements while the chunks come in (instead of hiding them via CSS), I believe @jakearchibald has done this before with some demo somewhere..

Reply:
Makes sense to me.  It's been a while, but I believe that @dvoytenko recommended that I take this approach - he said it was actually quite efficient.

The biggest difficulty with removing elements as they stream in is dealing with instances in which an element arrives in multiple chunks.  That said, that's already a TODO in this code, as it comes up when we try to detect when `<body>` has arrived.  Not insurmountable, of course.

Reply:
My recommendation was to simply reuse the existing `amp-shadow` class in the page's HTML. E.g. the AMP page would have something like this:

```
<style amp-custom>
header {
  /* styles for header in a standalone mode */
}
.amp-shadow header {
  display: none; /* do not show when inside a shadow doc */
}
</style>
```

This was no transformation is needed at all. Obviously, pre-transforming the doc would have its own benefits, but:

a. Transforming on client is very expensive, especially given chunks and streaming.
b. Transforming on server is fast, but requires a dependency on the server and complicates request protocol. Makes these pages harder to cache in AMP Cache.
c. Makes debugging hard.

A simple `.amp-shadow ... {}` selectors may seem simplistic, but they actually do not have any of these drawbacks. In particular:

a. Debugging is trivial: just toggle class in DevTools and see what it looks like.
b. No content transformation.
c. DOM overhead is likely very small. AMP does not load any `display: none` elements. 
d. The same AMP page is fully cacheable on AMP CDN (or any other CDN) in its single form and simply adopts to the way it's displayed.

Reply:
Thanks, Dima!  In this case, we're reusing AMPs created by the Guardian - so their CSS doesn't contain that handy `.amp-shadow` selector.  Thus, as the AMP streams in, I inject into it the CSS that hides the elements we don't want to see.  I don't use `.amp-shadow` because the CSS is only used when the AMP's living in the shadow root.

That said, in another PR, we just added a server that proxies Guardian article fetches.  This would let us inject CSS with `.amp-shadow` if we wanted to in the future.

Reply:
Yeah, if you do not own the document, obviously not much you can do - you have to resort to less efficient means. Please just make a comment to that effect. E.g. "have to do something very inefficient here. If I owned these docs, I'd just use amp-shadow class" or something like that.

Reply:
Makes sense.  Done!  And I'll also mention that in the blog post.